linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IntegerDivisions.scala

```diff
@@ -0,0 +1,699 @@
+            ClassType(LongImpl.RuntimeLongClass, nullable = true))
+
+        /* Use an explicit temp var to make sure the computation of `lo` can
+         * be dead-code-eliminated. For some reason, directly chaining the call
+         * to `hi()` leaves a record behind, which depends on `lo`.
+         */
+        val tDef = tempVarDef(mulhiTempName, multiplyFullCall)
+        Block(
+          tDef,
+          Apply(ApplyFlags.empty, tDef.ref, MethodIdent(LongImpl.hi), Nil)(IntType)
+        )
+      } else {
+        // ((x.toLong * y.toLong) >>> 32).toInt
+        UnaryOp(
+          UnaryOp.LongToInt,
+          BinaryOp(
+            BinaryOp.Long_>>>,
+            BinaryOp(
+              BinaryOp.Long_*,
+              LongLiteral(x.toLong),
+              UnaryOp(UnaryOp.IntToLong, y)
+            ),
+            IntLiteral(32)
+          )
+        )
+      }
+    }
+
+    def genMulUnsignedHi(x: Int, y: VarRef)(implicit pos: Position): Tree = {
+      // ((Integer.toUnsignedLong(x) * Integer.toUnsignedLong(y)) >>> 32).toInt
+      UnaryOp(
+        UnaryOp.LongToInt,
+        BinaryOp(
+          BinaryOp.Long_>>>,
+          BinaryOp(
+            BinaryOp.Long_*,
+            LongLiteral(Integer.toUnsignedLong(x)),
+            UnaryOp(UnaryOp.UnsignedIntToLong, y)
+          ),
+          IntLiteral(32)
+        )
+      )
+    }
+  }
+
+  implicit object LongIsUnsignedIntegral
+      extends UnsignedIntegral[Long]
+      with Numeric.LongIsIntegral with Ordering.LongOrdering {
+
+    val bitSize = 64
+
+    def isUnsignedPowerOf2OrZero(x: Long): Boolean =
+      (x & (x - 1L)) == 0L
+
+    def log2Exact(x: Long): Int =
+      63 - JLong.numberOfLeadingZeros(x)
+
+    def toUnsignedBigInt(x: Long): BigInt = {
+      val allButSignBit = BigInt(x & ~Long.MinValue)
+      if (x < 0L)
+        allButSignBit.setBit(63)
+      else
+        allButSignBit
+    }
+
+    def fromBigInt(x: BigInt): Long =
+      x.toLong
+
+    // IR-related operations
+
+    val tpe: Type = LongType
+
+    def literal(x: Long)(implicit pos: Position): Literal = LongLiteral(x)
+
+    // scalastyle:off disallow.space.before.token
+    val Op_+ : BinaryOp.Code = BinaryOp.Long_+
+    val Op_- : BinaryOp.Code = BinaryOp.Long_-
+    val Op_* : BinaryOp.Code = BinaryOp.Long_*
+    val Op_& : BinaryOp.Code = BinaryOp.Long_&
+    val Op_>>> : BinaryOp.Code = BinaryOp.Long_>>>
+    val Op_>> : BinaryOp.Code = BinaryOp.Long_>>
+    // scalastyle:on disallow.space.before.token
+
+    def genMulSignedHi(x: Long, y: VarRef, useRuntimeLong: Boolean)(
+        implicit pos: Position): Tree = {
+      /* Math.multiplyHigh(x, y)
+       * Unfortunately, we cannot directly call that method, because it won't
+       * be available when we link a javalib < 1.20. So we hard-code its IR.
+       * On the flip side, that allows the ElementaryInterpreter to easily
+       * handle this code, so it's not lost.
```

Comment:
```suggestion
       * handle this code (only basic operations and literals required).
```

Or maybe I don't understand what you mean with "it's not lost".